specs/WebMessageObjects.md

```diff
@@ -0,0 +1,176 @@
+# Background
+This is a proposal for a new API that will provide a framework for native representation of DOM
+objects from page content in WebView2 and its implementation for DOM File objects. The (specific ask
+from WebView2)[https://github.com/MicrosoftEdge/WebView2Feedback/issues/501] is to be able get to
+the paths for DOM file objects, which is not accessible to page content. We also have asks to
+expose other DOM objects, including iframes, <object> objects, etc. which will be added under this
+WebMessageObjects framework in future.
+
+# Conceptual pages (How To)
+
+WebMessageObjects are representations of DOM objects that can be passed via the [WebView2 WebMessage
+API](https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2webmessagereceivedeventargs)
+to the app. You can examine supported DOM objects using native reflections of the types.
+
+Currently the only supported DOM object type with this API is:
+- [File](https://developer.mozilla.org/en-US/docs/Web/API/File)
+
+Page content in WebView2 can pass objects to the app via the
+`chrome.webview.postMessageWithAdditionalObjects(string message, array<object> objects)` content API
+that takes in the array of such supported DOM objects or you can also use
+(ExecuteScript)[https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.winforms.webview2.executescriptasync]
+with same API. If an invalid or unsupported object is passed via this API, an exception will be
+thrown to content and the message will fail to post.
+
+On the WebMessageReceived event handler, the app will retrieve the native representation of objects
+via `AdditionalObjects` property and cast the passed objects to their native types. For example, the
+DOM File object will be exposed as a `CoreWebView2File` object, allowing the app to read the path of
+the passed DOM File object.
+
+# Examples
+## Read the File path of a DOM File
+
+Use this API with a DOM File object to be able to get the path of files dropped on WebView2. The
+HTML and JavaScript snippets are part of the page content code running inside WebView2.
+
+```html
+<!-- File upload location -->
+<input type="file" id="files" multiple />
+```
+
+```javascript
+const input = document.getElementById('files');
+input.addEventListener('change', function() {
+    const curFiles = input.files;
```

Comment:
Will this work? The Input element's files property is a FileList object which behaves array-like (it has a length property and you can do curFiles[idx] to get entries out of it), but is not actually an array object (doesn't have all the array methods like reverse, sort, and so on, its prototype chain doesn't involve Array). 

Just not sure if you're committing to supporting array-like parameters here or actual array parameters.

Reply:
Good catch. This works because in the WebIDL, we accept the additionalObjects as a sequence<any>. In the TS documentation, I'll use the ArrayLike<any> type to represent this.